pipeline/extract/chat.py

```diff
@@ -0,0 +1,105 @@
+from typing import List, Set
+
+import pandas as pd
+import prefect
+from prefect import task
+
+from pipeline.types import (
+    ChatData,
+    ChatMatchId,
+    Community,
+    Match,
+    MatchMetadata,
+    Message,
+    User,
+    UserId,
+)
+
+# TODO:
+# fix naming conventions
+
+# add extract_chat_data in imports in flows/matching.py
+# call your funciton in the flow after the other extract tasks
+
+# Later: Write a function to join user objects and chatdata using user.uid
+# Function to join chatdata and all messages relating to that chat
+
+
+@task
+def extract_recent_chatdata(db, community: Community) -> List[ChatData]:
+    logger = prefect.context.get("logger")
+    chat_ref = db.reference(f"matches/{community}")
+    user_chat_records = chat_ref.get()
+    if not user_chat_records:
+        logger.info(f"No user-chat records for commmnity: {community}")
+        return []
+    logger.info(f"Extracted {len(user_chat_records)} user-chat records.")
+    ids: Set[ChatMatchId] = set()
+    chatList: List[ChatData] = []
+
+    for key in user_chat_records.keys():
+        if user_chat_records[key]["id"] not in ids:
+
+            timestamp = user_chat_records[key]["release_timestamp"]
+            id = user_chat_records[key]["id"]
+            participants = user_chat_records[key]["participants"]
+            logger.info(f"participants: {participants}")
+            rtag = user_chat_records[key]["release_tag"]
+            chatdata_title = user_chat_records[key]["title"]
+            raw_metadata = user_chat_records[key].get("metadata", {})
+            parsed_md = MatchMetadata(**raw_metadata)
+
+            # run extract_message
+            ext_message = extract_message_data(
+                db, community=community, chatid=id
+            )
+            if ext_message:
+                logger.info(f"Extracted message: {ext_message}")
+
+            cd1 = ChatData(
+                release_timestamp=timestamp,
+                chat_match_id=id,
+                community_id="sds",
+                participants=participants,
```

Comment:
If you look at the type definition for `ChatData`, `participants` should be a dictionary of user IDs to `User` objects. But the database only stores the user IDs. That means you need to look up the `User` objects by ID.

There are a couple ways you could accomplish that:
1. Create a separate function that takes the `ChatData` that was extract and then adds in the `User` objects to `participants`
2. Pass in a list of `User` objects to this function, and then look them up by user ID to add to the `participants` dictionary, that means you take a new argument, and that list of users must have all the users you will need
3. In this extract function, also extract all the users from the community using the function that exists, and then add them to `participants`

I would not recommend #2 because it involves making extra database requests in the same function that are not necessary. #1 or #2 would be easier to test and understand.

@notedwin can give you advice on how to do this!

Reply:
A separate function was added to help with modifying the participants into showing User objects. This was then used for the `extract_recent_chatdata` function.